kube-controllers/pkg/controllers/loadbalancer/loadbalancer_controller.go

```diff
@@ -468,114 +457,150 @@ func (c *loadBalancerController) syncService(svcObj serviceObject) error {
 			}
 		}
 	} else {
 		// No annotations are specified, check that the IPs assigned aren't from Manual pool from earlier assignment
-		for ip := range c.allocationTracker.ipsByService[svcObj.namespacedName] {
+		for ip := range c.allocationTracker.ipsByService[svcObj] {
 			pool, err := c.poolForIP(ip)
 			if err != nil {
 				return err
 			}
 			if pool != nil {
 				if pool.Spec.AssignmentMode == api.Manual {
-					err = c.releaseIP(svcObj.namespacedName, ip)
+					err = c.releaseIP(svcObj, ip)
 					if err != nil {
 						log.WithError(err).Errorf("Failed to release IP for %s/%s", svc.Namespace, svc.Name)
 						return err
 					}
-					c.allocationTracker.releaseAddressFromService(svcObj.namespacedName, ip)
 				}
 			}
 		}
 	}
 
-	if IsCalicoManagedLoadBalancer(svc, c.cfg.AssignIPs) {
-		statusIPs = make(map[string]bool)
-		for _, ingress := range svc.Status.LoadBalancer.Ingress {
-			statusIPs[ingress.IP] = true
+	if c.needsIPsAssigned(svc, svcObj) {
+		_, err = c.assignIP(svc)
+		if err != nil {
+			log.WithError(err).Errorf("Failed to assign IP for %s/%s", svc.Namespace, svc.Name)
 		}
+	}
 
-		for ip := range c.allocationTracker.ipsByService[svcObj.namespacedName] {
-			if _, ok := statusIPs[ip]; !ok {
-				break
-			}
-			return nil
+	if c.needsStatusUpdate(svc, svcObj) {
+		err = c.updateServiceStatus(svc, svcObj)
+		if err != nil {
+			log.WithError(err).Errorf("Failed to update service status for %s/%s", svc.Namespace, svc.Name)
+			return err
 		}
+	}
 
-		_, err = c.assignIP(svc)
+	return nil
+}
 
-		if err != nil {
-			log.WithError(err).Errorf("Failed to assign IP for %s/%s", svc.Namespace, svc.Name)
-		}
+// needsIPsAssigned determines if service IPFamilyPolicy is requirement is fulfilled by number of assigned IPs in IPAM storage
+func (c *loadBalancerController) needsIPsAssigned(svc *v1.Service, svcObj serviceObject) bool {
+	switch *svc.Spec.IPFamilyPolicy {
+	case v1.IPFamilyPolicySingleStack:
+		return len(c.allocationTracker.ipsByService[svcObj]) == 1
+	case v1.IPFamilyPolicyRequireDualStack, v1.IPFamilyPolicyPreferDualStack:
+		return len(c.allocationTracker.ipsByService[svcObj]) == 2
+	default:
+		return true
+	}
+}
 
-		svcIngress := []v1.LoadBalancerIngress{}
-		for ip := range c.allocationTracker.ipsByService[svcObj.namespacedName] {
-			svcIngress = append(svcIngress, v1.LoadBalancerIngress{
-				IP: ip,
-			})
+// needsStatusUpdate checks if the service needs status update.
+// service needs status update if the IPs in our IPAM tracker do not match what's in the status of the service
+func (c *loadBalancerController) needsStatusUpdate(svc *v1.Service, svcObj serviceObject) bool {
+	if len(svc.Status.LoadBalancer.Ingress) != len(c.allocationTracker.ipsByService[svcObj]) {
+		return true
+	}
+
+	for _, ingress := range svc.Status.LoadBalancer.Ingress {
+		if _, ok := c.allocationTracker.ipsByService[svcObj][ingress.IP]; !ok {
+			break
 		}
-		svc.Status.LoadBalancer.Ingress = svcIngress
+		return false
```

Comment:
I don't think this is quite right. It's only going to check the first entry in the `Ingress` list before returning. This works correctly if the first entry in the status is not allocated, but doesn't work correctly when the first entry is allocated and the second is not (since we'll fall through and return `false` before ever executing the loop a second time). 

We need to check both that:
- All IPs that we own are present in the status
- All IPs that we own in the status are actually assigned in our IPAM DB